godot-core/src/task/futures.rs

```diff
@@ -308,8 +335,162 @@ impl<C: WithBaseField, R: ParamTuple + Sync + Send> IntoFuture for &TypedSignal<
     }
 }
 
+/// Convert a value into a type that is [`Send`] at compile-time while the value might not be.
+///
+/// This allows to turn any implementor into a type that is `Send`, but requires to also implement [`DynamicSend`] as well.
+/// The later trait will verify if a value can actually be sent between threads at runtime.
+pub trait IntoMaybeSend: Sealed {
+    type Target: Send + DynamicSend<Target = Self>;
+
+    fn into_maybe_send(self) -> Self::Target;
+}
+
+/// Check that it was safe to send the value to the current thread.
+///
+/// # Safety
+/// The implementor has to guarantee that `assert_safe_send` panics if the value has been sent between threads while being `!Send`.
+pub unsafe trait DynamicSend: Send + Sealed {
+    type Target;
+
+    fn extract_if_safe(self) -> Option<Self::Target>;
+}
+
+pub struct ThreadConfined<T> {
+    value: T,
+    thread_id: ThreadId,
+}
+
+unsafe impl<T> Send for ThreadConfined<T> {}
+unsafe impl<T> Sync for ThreadConfined<T> {}
+
+impl<T> ThreadConfined<T> {
+    pub(crate) fn new(value: T) -> Self {
+        Self {
+            value,
+            thread_id: std::thread::current().id(),
+        }
+    }
+
+    pub(crate) fn extract(self) -> Option<T> {
+        (self.thread_id == std::thread::current().id()).then_some(self.value)
+    }
```

Comment:
Can we remove this method, it's just duplicating `extract_if_safe`?

Having a smaller API surface makes it easier to prove that the `unsafe` guarantee of `DynamicSend` -- especially the one regarding public APIs -- is upheld.

Reply:
The `impl_dynamic_send` macro uses this function, so we don't have to duplicate the implementation.